base/intfuncs.jl

```diff
@@ -319,16 +319,19 @@ const HWNumber = Union{HWReal, Complex{<:HWReal}, Rational{<:HWReal}}
 @inline literal_pow(::typeof(^), x::HWNumber, ::Val{1}) = x
 @inline literal_pow(::typeof(^), x::HWNumber, ::Val{2}) = x*x
 @inline literal_pow(::typeof(^), x::HWNumber, ::Val{3}) = x*x*x
+@inline literal_pow(::typeof(^), x::HWNumber, ::Val{-1}) = inv(x)
+@inline literal_pow(::typeof(^), x::HWNumber, ::Val{-2}) = i=inv(x); i*i
+@inline literal_pow(::typeof(^), x::HWNumber, ::Val{-3}) = i=inv(x); i*i*i
 
 # don't use the inv(x) transformation here since float^p is slightly more accurate
 @inline literal_pow(::typeof(^), x::AbstractFloat, ::Val{p}) where {p} = x^p
 @inline literal_pow(::typeof(^), x::AbstractFloat, ::Val{-1}) = inv(x)
 
 # for other types, define x^-n as inv(x)^n so that negative literal powers can
 # be computed in a type-stable way even for e.g. integers.
-@inline function literal_pow(f::typeof(^), x, ::Val{p}) where {p}
+@inline function literal_pow(f::typeof(^), x, ::Val{p}) where {p<:Integer}
     if p < 0
-        literal_pow(^, inv(x), Val(-p))
+        f(Float64(x), p)
```

Comment:
I don't understand how you can convert to a `Float64` unconditionally here. Say you have:

```jl
struct Foo{T} <: Number
    x::T
end
Base.:(/)(f1::Foo, f2::Foo) = Foo(f1.x / f2.x)
Base.:(*)(f1::Foo, f2::Foo) = Foo(f1.x * f2.x)

f(x) = x^(-3)
x = Foo(3)
f(x)
```

Then it will try do `Float64(x)` which is not defined?

Reply:
Or just `x::Matrix`, which also takes this code path.

Reply:
so should add a method for `literal_pow(^, x::Integer, Val{P})`?

Reply:
Not an `Integer` in general, that has the same issue. Has to be something that only has types in Base. 

But I don't know what this is trying to fix. There is no corresponding rest for example. Why should it convert to a `Float64` in specifically the literal case?

Reply:
The reasoning is that `Float64^Integer` does a compensated inverse, so it is more accurate.

Reply:
I see you added `Integer` but that has the same issue as my earlier post in that `Float64(x)` might not make sense.